### DISCUSSION TITLE
Error mounting SDCARD in micropython v1.24.0 ESP32_GENERIC

### DISCUSSION THREAD

**Discussion by AledelaCV** [micropython/micropython#16125] [ESP32]
I have upgraded to version 1.23.0 to 1.24.0 encountering the error when mounting SDCard. The sample code in the previous version works without problem, but in the new version of micropython it generates error: 
```
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
OSError: [Errno 5] EIO
```

Here's the code I'm using:
```
import machine, os, vfs

# Slot 2 uses pins sck=18, cs=5, miso=19, mosi=23
sd = machine.SDCard(slot=2)
vfs.mount(sd, '/sd') # mount

print(os.listdir('/sd'))    # list directory contents
```

This code works perfectly in version 1.23.0

**Comment by shariltumin** [comment]
I can confirm that we have problems mounting devices from machine.SDCard() both for slot=1 and slot=2.

```
MicroPython v1.25.0-preview.4.g9ea8d2a03.dirty on 2024-10-29; ESP32-CAM OV2640 (KAKI5) with ESP32
>>>
>>> from machine import SDCard
>>> import vfs
>>> sd = SDCard(slot=1)
>>> sd.info()
(8053063680, 512)
>>> vfs.mount(sd, '/sd')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 5] EIO
>>>
```

The ```sd.info()``` works fine.

**Comment by shariltumin** [comment]
It seems that there is not much interest in the (non) working of machine.SDCard() for esp32. I've been trying to find out what is the reason why the machine.SDCard() is now broken. The code for machine_sdcard.c and extmod/vfs_* has been stable for 8 months, except for vfs_blockdev.c, which was changed about 2 months ago.

Now, when I replace the vfs_blockdev.c from the current release with the one from v1.22.0-preview.246.g4365edb81 and recompile, the machine.SDCard() seems to work as expected, at least with the following test script:

```
import esp
esp.osdebug(esp.LOG_VERBOSE)

import os, vfs
from machine import Pin, SDCard

sd = SDCard(slot=1, width=1)
print(sd.info())

vfs.mount(sd, '/sd')
print(os.listdir('/sd'))

``` 

Now, it seems that the introduction of mp_vfs_blockdev_call_rw() breaks the machine.SDCard(). I am not too knowledgeable to find out where is the real problem is. I hope Angus can find some time to look into this.

For now, we can use "sdcard.py" library found at 
https://github.com/mendenm/micropython-lib/tree/sdcard-fixes/micropython/drivers/storage/sdcard

For ESP32-CAM, please try this script:

```
from sdcard import SDCard
import os, vfs
from machine import Pin, SoftSPI

cs = Pin(13) # D03
#cs = Pin(12) # D02 - Err. OSError: [Errno 19] ENODEV: no SD card
#cs = Pin(4) # D01 - Err. OSError: [Errno 19] ENODEV: no SD card (Flash on!)

# MMS to SPI conversion ESP32-CAM sdcard
# miso=D00=2, sck=clk=14, mosi=CMD=15, cs=D03=13

spi = SoftSPI(sck=14, mosi=15, miso=2)
sd = SDCard(spi, cs)

vfs.mount(sd, '/sd')
print(os.listdir('/sd'))
```

**Comment by shariltumin** [comment]
The mp_vfs_blockdev_call_rw() calls from mp_vfs_blockdev_read() and mp_vfs_blockdev_write() return -22.  For them to work, they must both return 0.

The reason for -22 being returned is due to:

```
int i = MP_OBJ_SMALL_INT_VALUE(ret);
return i > 0 ? (-MP_EINVAL) : i;
```

since ```ret = mp_call_method_n_kw(n_args, 0, args)``` always returns 15

MP_EINVAL(Invalid argument) is 22, while
MP_ENOTBLK(Block device required) is 15.

If I set the return value of mp_vfs_blockdev_call_rw() in "extmod/vfs_blockdev.c" to 0 for both blockdev read and write (as of v1.22.0-preview.246.g4365edb81), everything works as normal for ESP32 (I have not checked for other boards).

If you are interested, you can find the actual files at https://github.com/shariltumin/bit-and-pieces/tree/main/SDCard%20Problem

**Comment by dpgeorge** [comment]
Thanks for the report and discussion here.  This definitely needs to be fixed, and will go in a 1.24.1 patch release.

**Reply by shariltumin** [reply]
Thanks Damien.

**Reply by AledelaCV** [reply]
I appreciate the interest shown in the subject. We will wait for the update to include the solution to the problem. thank you.
